include/libhal-arm-mcu/stm32f1/adc.hpp

```diff
@@ -0,0 +1,89 @@
+// Copyright 2024 Khalil Estell
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LIC`ENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <cstdint>
+
+#include <libhal/adc.hpp>
+#include <libhal/initializers.hpp>
+#include <libhal/units.hpp>
+
+namespace hal::stm32f1 {
+/**
+ * @brief Analog to digital converter
+ *
+ */
+class adc final : public hal::adc
+{
+public:
+  /// Channel specific information
+  struct channel
+  {
+    /// ADC port
+    std::uint8_t port;
+    /// ADC pin
+    std::uint8_t pin;
+    /// Channel data index
+    uint8_t index;
+  };
+
+  /**
+   * @brief Get a predefined adc channel
+   *
+   * - ADC channel 0 is PA0
+   * - ADC channel 1 is PA1
+   * - ADC channel 2 is PA2
+   * - ADC channel 3 is PA3
+   * - ADC channel 4 is PA4
+   * - ADC channel 5 is PA5
+   * - ADC channel 6 is PA6
+   * - ADC channel 7 is PA7
+   * - ADC channel 8 is PB0
+   * - ADC channel 9 is PB1
+   * - ADC channel 10 is PC0
+   * - ADC channel 11 is PC1
+   * - ADC channel 12 is PC2
+   * - ADC channel 13 is PC3
+   * - ADC channel 14 is PC4
+   * - ADC channel 15 is PC5
+   *
+   * @param p_channel - which adc channel to use
+   */
+  adc(hal::channel_param auto p_channel)
```

Comment:
So I previously used the "channel" to designate which ADC you are grabbing. But from a developers perspective, you I bet you will agree, Its harder to figure out which channel to use but its easier to determine which pin to use because that is what you are connecting to on your board/pcb, etc. 

So another strategy you can employ here, since each channel seems to have a 1-to-1 corrispondance to a specific pin, is it have an enum class of pins and pass those. Their "enum" value would be the channel number and you'd no longer need to have the channel object with 3-bytes (which turns into 4-bytes for alignment purposes).

```C++
enum class pins {
  pa0 = 0,
  pa1 = 1,
  // ...
  pc4 = 14,
  pc5 = 15,
};
```

If the set above is not all of the possible combinations, then the channel struct is still useful, but could be in its own constructor like you have below.